Tomviz conda recipe

cjh1 opened the pull request.

cjh1 committed:
Initial build recipe for Tomviz

cjh1 committed:
MNT: Re-rendered with conda-build 3.20.5, conda-smithy 3.8.5, and conda-forge-pinning 2020.12.02.10.35.15

mrakitin commented:
Power-cycled to trigger a CI build.

cjh1 commented:
Thanks @mrakitin

mrakitin commented:
My pleasure, @cjh1!

Is it supposed to produce different versions of the package for different Python versions?

cjh1 commented:
Nope, its fixed to one version currently (3.7)

mrakitin commented on recipe/meta.yaml:
```diff
@@ -1,36 +1,98 @@
 test:
   imports:
-    - tomviz
+    - paraview
```
Is it supposed to be `paraview` here?

mrakitin replied:
... the test is failing on importing it: https://dev.azure.com/nsls2forge/nsls2forge/_build/results?buildId=5742&view=logs&j=a70f640f-cc53-5cd3-6cdc-236a1aa90802&t=f5d15007-a01c-5ad8-c9ce-4d519d3b275f&l=433.

cjh1 replied:
Yes, for it just paraview. There isn't really a tomviz package to test import as its a Qt app. We can refine the testing later. The test is failing on Windows as we currently aren't building anything on Window. This just has the bare bones for Linux.

cjh1 committed:
Add git-lfs

Need by vtk-m

cjh1 committed:
Fix ParaView_DIR

cjh1 committed:
Remove hdf5 we get ours from VTK

cjh1 committed:
For now use tomviz repo

We need fixes that currently aren't in a release

cjh1 committed:
Add bld.bat for Windows

cjh1 committed:
Fix up CMake types

cjh1 committed:
Pin to Python 3.7.8

The latest 3.7.9 seems to have broken headers.

cjh1 committed:
Set PARAVIEW_CUSTOM_LIBRARY_SUFFIX

Set PARAVIEW_CUSTOM_LIBRARY_SUFFIX to avoid clashing with installed
ParaView. It would be nice to get the value/version from meta.yaml
but there currently isn't a nice way todo it.

mrakitin commented on recipe/meta.yaml:
```diff
@@ -1,36 +1,100 @@
+    - tbb-devel
   run:
-    - python
+    - python=3.7.8
```
It's better to avoid such hard pins. Better to build the package for multiple Python versions as needed. This is configured via `conda_build_config.yaml` file, as in other feedstocks.

cjh1 replied:
Tomviz currently only supports python 3.7, but you are saying set that in `conda_build_config.yaml`?

mrakitin replied:
Yes.

mrakitin reviewed (changes requested):
I think we are missing `recipe/conda_build_config.yaml` file. We use this file as an example: https://github.com/nsls-ii-forge/event-model-feedstock/blob/master/recipe/conda_build_config.yaml. This is to limit the channels to `nsls2forge` and `defaults`. We are currently not using `conda-forge` since the rest of the packages from `nsls2forge` were built to use `nsls2forge` and `defaults` channels exclusively (which are not 100% compatible with `conda-forge`).

cjh1 committed:
Add conda_build_config.yaml

mrakitin commented:
Rerender is needed after `conda_build_config.yaml` is changed.

cjh1 committed:
Relax python version

cjh1 commented:
@mrakitin OK, thanks I will rerender.

cjh1 committed:
MNT: Re-rendered with conda-build 3.20.5, conda-smithy 3.8.6, and conda-forge-pinning 2021.01.04.22.18.20

cjh1 committed:
Update version string to use master

Update to the form <version>.post<number of commits after that tag>

cjh1 commented:
@mrakitin Are the error on related setting `cdt_name` to `cos6` in `conda_build_config.yaml`

cjh1 commented:
@mrakitin Actually, I think I need to add `conda-forge` to  `channel_sources`.

cjh1 committed:
Add conda-forge to channel sources

cjh1 committed:
MNT: Re-rendered with conda-build 3.20.5, conda-smithy 3.8.6, and conda-forge-pinning 2021.01.04.22.18.20

mrakitin commented:
Let's try that. I believe there was an issue with the missing compiler v9 version. I think that the `defaults` channel is lagging, and only providing v7. Can we try that instead? Building locally with the https://github.com/nsls-ii-forge/tomviz-feedstock/blob/tomviz-build/build-locally.py script can help to resolve it faster. Alternatively, I sometimes just start a container and package there to debug more interactively.

cjh1 commented:
@mrakitin I think v7 would work, I'm not explicitly request v9 as far as I know.

cjh1 commented:
@mrakitin Thanks for the build local option.

cjh1 commented:
Local build shows it can find v9 from `conda-forge`

mrakitin commented:
> Local build shows it can find v9 from `conda-forge`

Can we try v7 from `defaults`, without `conda-forge` for now?

We do have a plan to switch to conda-forge for our distribution channel, but it needs efforts and time.

cjh1 commented:
> > Local build shows it can find v9 from `conda-forge`
> 
> Can we try v7 from `defaults`, without `conda-forge` for now?
> 
> We do have a plan to switch to conda-forge for our distribution channel, but it needs efforts and time.

Yes, I can try that, but there are other packages that it seems to need from `conda-forge` as  well, such as `git-lfs`. Where do I set the version?

cjh1 commented:
Looks like after adding `conda_build_config.yaml` CMake can't find OpenGL. I think I am using the right CDT ( mesa-libgl-devel ).

mrakitin commented:
> Yes, I can try that, but there are other packages that it seems to need from `conda-forge` as well, such as `git-lfs`. Where do I set the version?

It would be an edited entry, such as this one, in `conda_build_config.yaml`: https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/fe759728f24658d3031c3f71459cf774948ce389/recipe/conda_build_config.yaml#L7-L20.

The `defaults` channel provides `git-lfs`: https://anaconda.org/anaconda/git-lfs. But it's a bit older than in `conda-forge`. Do you need a newer version?

mrakitin commented:
> Looks like after adding `conda_build_config.yaml` CMake can't find OpenGL. I think I am using the right CDT ( mesa-libgl-devel ).

Hmm, the meta-packages are there: https://anaconda.org/search?q=mesa-libgl-devel. Maybe we need to set [these](https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/fe759728f24658d3031c3f71459cf774948ce389/recipe/conda_build_config.yaml#L115-L119) too?
```yaml
cdt_name:  # [linux]
  - cos6   # [linux64]
  - cos7   # [linux and aarch64]
  - cos7   # [linux and ppc64le]
  - cos7   # [linux and armv7l]
```

cjh1 commented:
> > Yes, I can try that, but there are other packages that it seems to need from `conda-forge` as well, such as `git-lfs`. Where do I set the version?
> 
> It would be an edited entry, such as this one, in `conda_build_config.yaml`: https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/fe759728f24658d3031c3f71459cf774948ce389/recipe/conda_build_config.yaml#L7-L20.
> 
> The `defaults` channel provides `git-lfs`: https://anaconda.org/anaconda/git-lfs. But it's a bit older than in `conda-forge`. Do you need a newer version?

No, but if I don't include `conda-forge`, conda can't find the package for windows from the default channels.

cjh1 commented:
@mrakitin So I have experimented locally with using just the `defaults` and `nsls2forge` channel and I am missing the following deps:

```
  - gl2ps
  - libnetcdf=4.7.4
  - qt=5.12
  - cgns
  - glew=2.1
  - pugixml
```

cjh1 committed:
Add macos ParaView fix

cjh1 committed:
Use v7 C/C++ compiler

cjh1 committed:
Relax python version for host

cjh1 committed:
Add yum_requirements.txt

cjh1 committed:
MNT: Re-rendered with conda-build 3.20.5, conda-smithy 3.8.6, and conda-forge-pinning 2021.01.04.22.18.20

mrakitin commented:
> @mrakitin So I have experimented locally with using just the `defaults` and `nsls2forge` channel and I am missing the following deps:
> 
> ```
>   - gl2ps
>   - libnetcdf=4.7.4
>   - qt=5.12
>   - cgns
>   - glew=2.1
>   - pugixml
> ```

Are they build dependencies?

Can we use qt=5.9 - the "latest" version `defaults` provides? I know it's a bit outdated, but that's what we deploy on the floor, and have other Qt apps running against that version.

mrakitin commented:
So, here is the status:
  - `gl2ps` - exists in conda-forge, we can build it in our channel (I hope) based on https://github.com/conda-forge/gl2ps-feedstock/blob/master/recipe/meta.yaml
  - `libnetcdf=4.7.4` - `defaults` has v4.7.3. Will it work? Or do you need this exact pin?
  - `qt=5.12` - as mentioned before, `defaults` hasn't caught up with `conda-forge`, we may want to downgrade the pin (use `qt >=5.9`)
  - `cgns` - can be built based on https://github.com/conda-forge/cgns-feedstock/blob/master/recipe/meta.yaml.
  - `glew=2.1` - can be built based on https://github.com/conda-forge/glew-feedstock/blob/master/recipe/meta.yaml, but it's very heavy, and may require more dependency packages to be built
  - `pugixml` - can be built with https://github.com/conda-forge/pugixml-feedstock/blob/master/recipe/meta.yaml.

mrakitin commented:
I have a thought - if these dependencies are _build_ only, we can include them from `conda-forge`, but they should not be in the list of _run_ dependencies.

mrakitin commented on recipe/meta.yaml:
```diff
@@ -60,7 +60,7 @@ requirements:
     - {{ cdt('libuuid-devel') }}         # [linux]
   host:
-    - python=3.7.8
+    - python=3.7
```
Python's version will be picked from `conda_build_config.yaml`.
```suggestion
    - python
```

cjh1 replied:
OK, I can fix that

cjh1 commented:
> I have a thought - if these dependencies are _build_ only, we can include them from `conda-forge`, but they should not be in the list of _run_ dependencies.

Not sure what the difference between `host` and `run` is ( they are currently under `host` ). They are build dependencies, but the runtime shared libraries are needed at runtime, however they should be included in the package by the CMake build I think.

cjh1 committed:
Relax host python version

cjh1 commented:
@mrakitin I think I can probably remove the need to `libnetcdf`

cjh1 commented:
@mrakitin Looks like using Qt from nsls2forge causes a link error, see latest build.

mrakitin commented:
> @mrakitin Looks like using Qt from nsls2forge causes a link error, see latest build.

Well, strictly speaking, qt comes from `defaults`, not our channel.

Looks like some symbols are missing in that version:
```
$BUILD_PREFIX/bin/../lib/gcc/x86_64-conda_cos6-linux-gnu/7.3.0/../../../../x86_64-conda_cos6-linux-gnu/bin/ld: $BUILD_PREFIX/x86_64-conda_cos6-linux-gnu/sysroot/usr/lib64/libSM.so: undefined reference to `uuid_unparse_lower@UUID_1.0'
$BUILD_PREFIX/bin/../lib/gcc/x86_64-conda_cos6-linux-gnu/7.3.0/../../../../x86_64-conda_cos6-linux-gnu/bin/ld: $BUILD_PREFIX/x86_64-conda_cos6-linux-gnu/sysroot/usr/lib64/libSM.so: undefined reference to `uuid_generate@UUID_1.0'
```

Maybe the `uuid` library is missing? https://man7.org/linux/man-pages/man3/uuid_unparse.3.html

cjh1 commented:
@mrakitin Sure, Just so I am clear is it a hard requirement that we don't pull anything from `conda-forge`?

mrakitin commented:
For such big packages like TomViz, probably we need to have bits from conda-forge.

We have 2 options here:

1) Use `conda-forge` in the channels list & add the `collection=2021C1.0` metapackage here as a run dependency (it comes from `nsls2forge`) and see if the packages are compatible.

2) I should have recommended starting from https://github.com/conda-forge/staged-recipes. There is a bigger community there, and experienced people who package a lot of C/C++ packages. That is a relatively rare exercise here for this channel. We did have great success with packaging `hkl` (in CF first, and then as a copy here in this org) - https://github.com/conda-forge/staged-recipes/pull/10819.

I think we can start from (1), and depending on the results, figure out what to do next. Does it seem OK?

cjh1 commented:
@mrakitin That sounds good to me. So I just replace the `host` deps with `collection=2021C1.0`?

mrakitin commented:
> @mrakitin That sounds good to me. So I just replace the `host` deps with `collection=2021C1.0`?

I meant to add an extra line in the `run` dependencies to include the metapackage.

mrakitin commented on recipe/meta.yaml:
```diff
@@ -1,36 +1,100 @@
+    - gl2ps
+    - tbb-devel
   run:
     - python
```
```suggestion
    - python
    - collection=2021C1.0
```

cjh1 replied:
Ok, thanks

cjh1 replied:
That didn't seem to go so well!

mrakitin replied:
That was my concern. I am seeing a lot of conflicts in the build log, e.g., numpy packaged with openblas (in `conda-forge`) vs. mkl (in `defaults`), qt 5.12, etc. That's why I recommended sticking with `nsls2forge` and `defaults`. Or we should completely switch to `conda-forge`, but it's not a fast process, as all 200+ packages need to be replicated in the `conda-forge` channel (or at least a non-overlapping subset, which is still a large amount).

cjh1 committed:
Add metapackage dependency

mrakitin commented:
I think at this point it would make sense to open a PR in conda-forge's staged-recipes repo, to create a feedstock there, using the `conda-forge` channel only. We do maintain a few core libraries there (bluesky, ophyd, databroker, suitcases), which may be enough. That will have to be deployed in a separate conda environment, distinct from collection-2021-1.0 for now, until we replicate all required packages in CF.

cc @cryos, @stuartcampbell

cjh1 commented:
I can also loop back to trying to get things working with `nsls2forge` and `defaults` if that is still the preferred approach, it may just burn a little time.

mrakitin commented:
> I can also loop back to trying to get things working with `nsls2forge` and `defaults` if that is still the preferred approach, it may just burn a little time.

I think it may be a faster approach than having all feedstocks replicated, so I like the idea. I can help with a few missing dependencies.

cjh1 commented:
OK, sound good, I think we might be able to pare down the deps further, I will see what I can do.

cryos commented:
I have a high level question of whether we can put Tomviz into an analysis environment that can use the more recent/complete dependencies available in the conda-forge channel. It seems to me that Qt 5.9 is pretty ancient at this point, glew is pretty common as a dependency, gl2ps maybe a little less so. It seems that @mrakitin is thinking in a similar vein above too, this is an analysis tool that doesn't necessarily need to be contained. Maybe we can talk more this afternoon at the DAMA meeting.

cjh1 commented:
Any progress on building the deps to push this PR forward?

mrakitin commented:
@cjh1, no, unfortunately, I didn't have time to tackle this. What's our current list?

mrakitin commented:
> OK, sound good, I think we might be able to pare down the deps further, I will see what I can do.

@cjh1, I was waiting for the updated list, did not think the ball is in my court yet. Please provide a list of dependencies we need, I'll start building them this week.

cjh1 commented:
```
  - gl2ps
  - qt=5.12
  - glew
```

mrakitin commented:
I'll see how much work would it be to build them, and if it's worth the pain (they may bring more dependencies, which may bring more, ...)

I do think it makes sense to create a conda-forge feedstock meanwhile, so we have a backup option to install from into a separate analysis environment.

cjh1 commented:
@mrakitin OK, I can make a start on a conda-forge feedstock.

cjh1 commented:
Here is the start of a conda-forge feedstock: https://github.com/conda-forge/staged-recipes/pull/13734
